partition/grubenv/grubenv.go

```diff
@@ -0,0 +1,118 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package grubenv
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+// FIXME: support for escaping (embedded \n in grubenv) missing
+type Grubenv struct {
+	env      map[string]string
+	ordering []string
+
+	path string
+}
+
+func NewGrubenv(path string) *Grubenv {
+	return &Grubenv{
+		env:  make(map[string]string),
+		path: path,
+	}
+}
+
+func (g *Grubenv) Getenv(name string) string {
+	return g.env[name]
+}
+
+func (g *Grubenv) Setenv(key, value string) {
+	for i, k := range g.ordering {
+		if k == key {
+			g.ordering = append(g.ordering[:i], g.ordering[i+1:]...)
```

Comment:
Is there a benefit in changing the order on every Set? Feels more comfortable to have them stable.